### PR TITLE
Advanced Next-Level Flavours And Skins For Mastodon™

### PR DESCRIPTION
**Pull request by marrus-sh**
> (Note 1,655 of those added lines are the win95 theme I didn't change things *that* much)

This is the more robust themeing system suggested by @ekiru, or as best as I could interpret it. Probably don't merge until I've had time to document since I know a lot of folks running glitchsoc are interested in custom themes and this is a breaking change (see below). But, I wanted to get eyes on it early. It's late and i'm tired so there might still be some problems/bugs.

This was the old theme format (only affected web view), which should be ignored after this change:

```yaml
pack: home.js
```

This is the new theme format (can affect all sorts of pages):
```yaml
pack:
  common: # or home, about, settings, etc…
    filename: common.js
    stylesheet: true
  home: home.js # You can specify just the filename if you don't need additional options

fallback: glitch # Uses glitch packs where unspecified
```

With the exception of locales (which aren't yet integrated into themeing), **_all_** of glitchsoc's assets are now loaded this way, *including* core stuff (see `app/javascript/core/theme.yml`). This hopefully will fix a lot of our prior problems involving static pages, including profile metadata or things just not being styled correctly.

This PR includes the `win95` theme from our friends at cybre, although it will still need some minor tweaking before it's completely compatible with glitch.

### CODE REVIEW

**Comment by marrus-sh** [comment]
Apparently webpack still needs to be configured to ignore the old syntax lol (it's trying to break `index.js` up into it's component parts from the looks of it)

File: app/javascript/themes/win95/theme.yml [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+pack:
+  common:
+    filename: index.js
+    stylesheet: true
+  #  All unspecified packs will inherit from glitch.
+
+#  By default, the glitch preloads will also be used here. You can
+#  disable them by setting `preload` to `null`.
+
+#      preload:
```
Comment (marrus-sh) [inline app/javascript/themes/win95/theme.yml]: This comment should be removed, preloads are now specified per-pack

File: app/views/layouts/embedded.html.haml [hunk 2]
```diff
@@ -4,10 +4,9 @@
     %meta{ charset: 'utf-8' }/
     %meta{ name: 'robots', content: 'noindex' }/
 
-    = stylesheet_pack_tag 'common', media: 'all'
     = javascript_pack_tag 'common', integrity: true, crossorigin: 'anonymous'
-    = stylesheet_pack_tag 'application', integrity: true, media: 'all'
+    = javascript_pack_tag 'embed', integrity: true, crossorigin: 'anonymous'
```
Comment (marrus-sh) [inline app/views/layouts/embedded.html.haml]: these javascript pack tags need to be removed

**Comment by marrus-sh** [comment]
Skins support! You can now specify skins for themes. A skin named `tooty` for the `rooty` theme can be specified in `app/javascript/skins/rooty/tooty`, for example. This can be either

- A folder, in which case the (s)css files inside should be named according to the pack whose styles they are intended to replace (eg, `home.scss`, `common.scss`), or
- A stylesheet file, in which case the pack is assumed to be `common`.

`win95` has been migrated over as a vanilla theme, so you can check that out to see how it looks. It's at `app/javascript/skins/vanilla/win95.scss`!

**Comment by marrus-sh** [comment]
There's more functionality i'd like to implement before merge but the code here should be functional at this point 🙏🏻

Things still to come:
- Testing support?? (This might be a separate PR)
- Renaming "themes" to "flavours" to prevent confusion with upstream themes
- If the pack file specified is a (s)css file, it'd be nice if `stylesheet` defaulted to `true` in `theme.yml`
- Update the docs!!

**Comment by marrus-sh** [comment]
Apparently `.match?` was added in Ruby 2.4 so I'll have to change that for Ruby 2.3 support. `set_path` in the registrations controller is a typo for `set_pack`.

**Comment by marrus-sh** [comment]
I haven't tested it yet and it's possible that renaming everything Broke Something Somewhere but @beatrix-bitrot I daresay this is about ready for some eyes 👀

**Comment by yipdw** [comment]
I fixed a merge conflict with `master`; will let the CI build run.  If that passes I can do a check over the Ruby code.

**Comment by marrus-sh** [comment]
I fixed the last of the problems with the renaming I think.

I don't understand why the deletes test is failing—I certainly haven't touched it……… 😫

**Comment by beatrix-bitrot** [comment]
bravo you two~

**Pull request merged**